File: timetable/timetable.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>
#include <fstream>
#include <sstream>
#include <ctime>
#include <iomanip>
#include <random>
#include "json.hpp"

using namespace std;
using json = nlohmann::json;

const int DAYS = 5;
const int SLOTS_PER_DAY = 6;
const vector<string> DAY_NAMES = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday"};
const vector<string> TIME_SLOTS = {"9:00-10:00", "10:00-11:00", "11:00-12:00", "12:00-1:00", "1:00-2:00", "2:00-3:00"};

struct Class {
    string name;
    int frequency_per_week;
    bool is_lab;
    int duration;
    bool is_combined;
    vector<string> professors;
};

struct Section {
    string name;
};

int num_sections;
vector<Section> sections;
vector<Class> classes;
vector<Class> combined_classes;

map<string, vector<vector<string>>> timetable;
map<string, map<string, string>> section_professors;

mt19937 rng;

bool canFit(const vector<string>& day, int start, int duration) {
    if (start + duration > SLOTS_PER_DAY) return false;
    for (int i = start; i < start + duration; ++i) {
        if (!day[i].empty()) return false;
    }
    return true;
}

bool isClassScheduledOnDay(const vector<string>& day, const string& className) {
    for (const string& slot : day) {
        if (!slot.empty() && slot.find(className) != string::npos) {
            return true;
        }
    }
    return false;
}

string selectRandomProfessor(const vector<string>& professors) {
    if (professors.empty()) return "TBD";
    uniform_int_distribution<int> dist(0, professors.size() - 1);
    return professors[dist(rng)];
}

void preAssignProfessors() {
    cout << "Pre-assigning professors..." << endl;
    for (const auto& sec : sections) {
        string secName = sec.name;
        for (const auto& c : classes) {
            section_professors[secName][c.name] = selectRandomProfessor(c.professors);
            cout << "  Section " << secName << ": " << c.name << " -> " << section_professors[secName][c.name] << endl;
        }
        for (const auto& c : combined_classes) {
            section_professors[secName][c.name] = selectRandomProfessor(c.professors);
            cout << "  Section " << secName << ": " << c.name << " (combined) -> " << section_professors[secName][c.name] << endl;
        }
    }
}

void loadInputFromJson(const string& filename) {
    ifstream inFile(filename);
    if (!inFile.is_open()) {
        cerr << "Error: Failed to open " << filename << endl;
        exit(1);
    }

    json j;
    try {
        inFile >> j;
    } catch (const json::exception& e) {
        cerr << "Error parsing JSON: " << e.what() << endl;
        exit(1);
    }

    if (!j.contains("numSections") || !j.contains("courses")) {
        cerr << "Error: Missing required fields in input.json" << endl;
        exit(1);
    }

    num_sections = j["numSections"];
    
    // Create sections
    sections.clear();
    for (int i = 0; i < num_sections; ++i) {
        Section s;
        s.name = string(1, 'A' + i);
        sections.push_back(s);
    }

    // Load courses
    classes.clear();
    for (const auto& item : j["courses"]) {
        Class c;
        c.name = item["name"];
        c.frequency_per_week = item["frequency"];
        c.duration = item["duration"];
        c.is_lab = (c.duration == 2);
        c.is_combined = false;
        if (item.contains("professors") && item["professors"].is_array()) {
            c.professors = item["professors"].get<vector<string>>();
        }
        classes.push_back(c);
        cout << "Loaded class: " << c.name << " (freq: " << c.frequency_per_week << ", duration: " << c.duration << ")" << endl;
    }

    // Load combined classes
    combined_classes.clear();
    if (j.contains("combinedClasses") && j["combinedClasses"].is_array()) {
        for (const auto& item : j["combinedClasses"]) {
            Class c;
            c.name = item["name"];
            c.frequency_per_week = item["frequency"];
            c.duration = item["duration"];
            c.is_lab = (c.duration == 2);
            c.is_combined = true;
            if (item.contains("professors") && item["professors"].is_array()) {
                c.professors = item["professors"].get<vector<string>>();
            }
            combined_classes.push_back(c);
            cout << "Loaded combined class: " << c.name << " (freq: " << c.frequency_per_week << ", duration: " << c.duration << ")" << endl;
        }
    }

    cout << "Successfully loaded " << num_sections << " sections, " 
         << classes.size() << " classes, " << combined_classes.size() << " combined classes." << endl;
}

void initTimetables() {
    timetable.clear();
    for (const auto& sec : sections) {
        timetable[sec.name] = vector<vector<string>>(DAYS, vector<string>(SLOTS_PER_DAY, ""));
    }
    cout << "Initialized timetables for " << sections.size() << " sections" << endl;
}

vector<int> getLabSlots() {
    return {0, 2, 4}; // Start slots for 2-hour labs
}

void assignCombinedClasses() {
    cout << "Assigning " << combined_classes.size() << " combined classes..." << endl;
    for (const auto& c : combined_classes) {
        cout << "Processing combined class: " << c.name << endl;
        for (int freq = 0; freq < c.frequency_per_week; ++freq) {
            bool placed = false;
            cout << "  Attempting to place occurrence " << (freq + 1) << "/" << c.frequency_per_week << endl;
            
            for (int d = 0; d < DAYS && !placed; ++d) {
                // Check if already scheduled on this day
                if (isClassScheduledOnDay(timetable[sections[0].name][d], c.name)) {
                    cout << "    Day " << d << ": Already scheduled" << endl;
                    continue;
                }

                vector<int> slots_to_try = (c.duration == 2) ? getLabSlots() : vector<int>{0, 1, 2, 3, 4, 5};

                for (int t : slots_to_try) {
                    if (t + c.duration > SLOTS_PER_DAY) continue;

                    bool all_free = true;
                    for (const auto& sec : sections) {
                        if (!canFit(timetable[sec.name][d], t, c.duration)) {
                            all_free = false;
                            break;
                        }
                    }

                    if (all_free) {
                        for (const auto& sec : sections) {
                            string prof = section_professors[sec.name][c.name];
                            for (int k = 0; k < c.duration; ++k) {
                                timetable[sec.name][d][t + k] = c.name + " (" + prof + ")";
                            }
                        }
                        placed = true;
                        cout << "    Placed on " << DAY_NAMES[d] << " at slot " << t << endl;
                        break;
                    }
                }
            }
            if (!placed) {
                cout << "    WARNING: Could not place occurrence " << (freq + 1) << endl;
            }
        }
    }
}

void assignIndividualClasses() {
    cout << "Assigning individual classes..." << endl;
    
    for (const auto& sec : sections) {
        cout << "Processing section " << sec.name << endl;
        auto& grid = timetable[sec.name];
        
        // Sort classes: labs first, then by frequency (descending)
        vector<Class> section_classes = classes;
        sort(section_classes.begin(), section_classes.end(), [](const Class& a, const Class& b) {
            if (a.is_lab != b.is_lab) return a.is_lab > b.is_lab;
            return a.frequency_per_week > b.frequency_per_week;
        });

        for (const auto& c : section_classes) {
            cout << "  Assigning " << c.name << " (freq: " << c.frequency_per_week 
                 << ", duration: " << c.duration << ")" << endl;
            
            for (int freq = 0; freq < c.frequency_per_week; ++freq) {
                bool placed = false;
                cout << "    Placing occurrence " << (freq + 1) << "/" << c.frequency_per_week << endl;
                
                for (int d = 0; d < DAYS && !placed; ++d) {
                    // Skip if already scheduled on this day
                    if (isClassScheduledOnDay(grid[d], c.name)) {
                        continue;
                    }

                    if (c.duration == 2) {
                        // Lab scheduling
                        vector<int> lab_slots = getLabSlots();
                        for (int t : lab_slots) {
                            if (canFit(grid[d], t, c.duration)) {
                                string prof = section_professors[sec.name][c.name];
                                for (int k = 0; k < c.duration; ++k) {
                                    grid[d][t + k] = c.name + " (" + prof + ")";
                                }
                                placed = true;
                                cout << "      Placed lab on " << DAY_NAMES[d] << " at slots " << t << "-" << (t+1) << endl;
                                break;
                            }
                        }
                    } else {
                        // Regular class scheduling
                        for (int t = 0; t < SLOTS_PER_DAY; ++t) {
                            if (canFit(grid[d], t, c.duration)) {
                                string prof = section_professors[sec.name][c.name];
                                grid[d][t] = c.name + " (" + prof + ")";
                                placed = true;
                                cout << "      Placed on " << DAY_NAMES[d] << " at slot " << t << " (" << TIME_SLOTS[t] << ")" << endl;
                                break;
                            }
                        }
                    }
                    if (placed) break;
                }
                if (!placed) {
                    cout << "      WARNING: Could not place occurrence " << (freq + 1) << endl;
                }
            }
        }
    }
}

void printTimetable() {
    cout << "\n=== Generated Timetables ===" << endl;
    for (const auto& sec : sections) {
        cout << "\nSection " << sec.name << ":" << endl;
        cout << "Time Slot    ";
        for (int d = 0; d < DAYS; ++d) {
            cout << setw(20) << DAY_NAMES[d];
        }
        cout << endl;
        
        for (int t = 0; t < SLOTS_PER_DAY; ++t) {
            cout << setw(12) << TIME_SLOTS[t] << " ";
            for (int d = 0; d < DAYS; ++d) {
                string slot = timetable[sec.name][d][t];
                if (slot.empty()) slot = "FREE";
                cout << setw(20) << slot.substr(0, 19);
            }
            cout << endl;
        }
    }
}

void saveOutputToJson(const string& filename) {
    json output;

    // Add metadata
    output["metadata"]["generatedAt"] = time(0);
    output["metadata"]["numSections"] = num_sections;
    output["metadata"]["daysPerWeek"] = DAYS;
    output["metadata"]["slotsPerDay"] = SLOTS_PER_DAY;
    output["metadata"]["dayNames"] = DAY_NAMES;
    output["metadata"]["timeSlots"] = TIME_SLOTS;

    // Add timetables - ensure no nulls
    for (const auto& sec : sections) {
        json timetableArray;
        for (int d = 0; d < DAYS; ++d) {
            json day;
            for (int t = 0; t < SLOTS_PER_DAY; ++t) {
                string slot = timetable[sec.name][d][t];
                // Replace empty slots with "FREE" to avoid nulls
                day.push_back(slot.empty() ? "FREE" : slot);
            }
            timetableArray.push_back(day);
        }
        output["sections"][sec.name] = timetableArray;
    }

    try {
        ofstream outFile(filename);
        if (!outFile.is_open()) {
            cerr << "Error: Could not create " << filename << endl;
            return;
        }
        outFile << output.dump(4);
        cout << "\nTimetable successfully saved to " << filename << "!" << endl;
    } catch (const exception& e) {
        cerr << "Error saving output: " << e.what() << endl;
    }
}

int main() {
    rng.seed(time(0));

    cout << "=== Timetable Generator ===" << endl;
    cout << "Reading input from input.json..." << endl;
    
    loadInputFromJson("input.json");
    initTimetables();
    preAssignProfessors();
    
    if (!combined_classes.empty()) {
        assignCombinedClasses();
    }
    
    assignIndividualClasses();
    
    printTimetable();
    saveOutputToJson("output.json");
    
    cout << "\n=== Generation Complete ===" << endl;
    return 0;
}

File: timetable_final_SEE/timetable.py
import random
import time

DAYS = 5
SLOTS_PER_DAY = 6

class Class:
    def __init__(self, name, frequency_per_week, duration, is_combined, professors):
        self.name = name
        self.frequency_per_week = frequency_per_week
        self.is_lab = duration == 2
        self.duration = duration
        self.is_combined = is_combined
        self.professors = professors

class Section:
    def __init__(self, name, semester):
        self.name = name
        self.semester = semester

sections = []
classes = {}  # keyed by semester
combined_classes = {}  # keyed by semester
timetable = {}  # keyed by section name
section_professors = {}  # keyed by section name and class name
professor_schedule = {}  # keyed by professor and day -> list of occupied slots
random.seed(time.time())

def can_fit(day, start, duration):
    return all(day[i] == "" for i in range(start, start + duration) if i < SLOTS_PER_DAY)

def is_class_scheduled_on_day(day, class_name):
    return any(class_name in slot for slot in day)

def is_professor_available(prof, day, start, duration):
    for i in range(start, start + duration):
        if i in professor_schedule.get(prof, {}).get(day, set()):
            return False
    return True

def occupy_professor(prof, day, start, duration):
    if prof not in professor_schedule:
        professor_schedule[prof] = {}
    if day not in professor_schedule[prof]:
        professor_schedule[prof][day] = set()
    for i in range(start, start + duration):
        professor_schedule[prof][day].add(i)

def select_random_professor(professors):
    return random.choice(professors)

def pre_assign_professors():
    # Initialize for each section
    for section in sections:
        semester = section.semester
        section_professors[section.name] = {}

    # Assign professors for individual classes
    for semester in classes:
        for c in classes[semester]:
            relevant_sections = [s for s in sections if s.semester == semester]
            shuffled_profs = c.professors.copy()
            random.shuffle(shuffled_profs)
            for i, sec in enumerate(relevant_sections):
                assigned_prof = shuffled_profs[i % len(shuffled_profs)]
                section_professors[sec.name][c.name] = assigned_prof

    # Assign professors for combined classes
    for semester in combined_classes:
        for c in combined_classes[semester]:
            relevant_sections = [s for s in sections if s.semester == semester]
            shuffled_profs = c.professors.copy()
            random.shuffle(shuffled_profs)
            for i, sec in enumerate(relevant_sections):
                assigned_prof = shuffled_profs[i % len(shuffled_profs)]
                section_professors[sec.name][c.name] = assigned_prof

def init_timetables():
    for section in sections:
        timetable[section.name] = [["" for _ in range(SLOTS_PER_DAY)] for _ in range(DAYS)]

def get_lab_slots():
    return [0, 2, 4]

def assign_combined_classes():
    for semester in combined_classes:
        for c in combined_classes[semester]:
            relevant_sections = [s for s in sections if s.semester == semester]
            for _ in range(c.frequency_per_week):
                placed = False
                for d in range(DAYS):
                    if placed: break
                    if is_class_scheduled_on_day(timetable[relevant_sections[0].name][d], c.name):
                        continue
                    slots_to_try = get_lab_slots() if c.duration == 2 else list(range(SLOTS_PER_DAY))
                    for t in slots_to_try:
                        if t + c.duration > SLOTS_PER_DAY:
                            continue
                        if all(can_fit(timetable[sec.name][d], t, c.duration) for sec in relevant_sections):
                            prof_valid = True
                            for sec in relevant_sections:
                                prof = section_professors[sec.name][c.name]
                                if not is_professor_available(prof, d, t, c.duration):
                                    prof_valid = False
                                    break
                            if not prof_valid:
                                continue
                            for sec in relevant_sections:
                                prof = section_professors[sec.name][c.name]
                                for k in range(c.duration):
                                    timetable[sec.name][d][t + k] = f"{c.name} ({prof})"
                                occupy_professor(prof, d, t, c.duration)
                            placed = True
                            break

def assign_individual_classes():
    for section in sections:
        semester = section.semester
        grid = timetable[section.name]
        section_classes = sorted(classes[semester], key=lambda x: (not x.is_lab, -x.frequency_per_week))
        for c in section_classes:
            for _ in range(c.frequency_per_week):
                placed = False
                day_order = random.sample(range(DAYS), DAYS)
                for day_idx in day_order:
                    if placed: break
                    if is_class_scheduled_on_day(grid[day_idx], c.name):
                        continue
                    slots_to_try = get_lab_slots() if c.duration == 2 else list(range(SLOTS_PER_DAY))
                    random.shuffle(slots_to_try)
                    for t in slots_to_try:
                        if can_fit(grid[day_idx], t, c.duration):
                            prof = section_professors[section.name][c.name]
                            if is_professor_available(prof, day_idx, t, c.duration):
                                for k in range(c.duration):
                                    grid[day_idx][t + k] = f"{c.name} ({prof})"
                                occupy_professor(prof, day_idx, t, c.duration)
                                placed = True
                                break

def load_input_dict(data):
    for sem_data in data["semesters"]:
        sem = sem_data["semester"]
        for i in range(sem_data["numSections"]):
            sections.append(Section(f"{sem}_Section_{i+1}", sem))

        classes[sem] = []
        for c in sem_data["courses"]:
            classes[sem].append(Class(c["name"], c["frequency"], c["duration"], False, c["professors"]))

        combined_classes[sem] = []
        for c in sem_data["combinedClasses"]:
            combined_classes[sem].append(Class(c["name"], c["frequency"], c["duration"], True, c["professors"]))

def get_output_dict():
    out = {}
    for section, grid in timetable.items():
        out[section] = [day for day in grid]
    return out
